src/langserver/debugger/auxtools.rs

```diff
@@ -0,0 +1,432 @@
+        let (responses_sender, responses_receiver) = mpsc::channel();
+        let last_error = Arc::new(RwLock::new("".to_owned()));
+
+        let thread = {
+            let seq = seq.clone();
+            let last_error = last_error.clone();
+            AuxtoolsThread {
+                seq,
+                responses: responses_sender,
+                last_error: last_error,
+            }.spawn_listener(listener, connection_sender)
+        };
+
+        Ok((port, Auxtools {
+            seq,
+            responses: responses_receiver,
+            _thread: thread,
+            stream: StreamState::Waiting(connection_receiver),
+            last_error,
+        }))
+    }
+
+    fn read_response_or_disconnect(&mut self) -> Result<Response, Box<dyn std::error::Error>> {
+        match self.responses.recv_timeout(std::time::Duration::from_secs(5)) {
+            Ok(response) => Ok(response),
+            Err(_) => {
+                self.disconnect();
+                Err(Box::new(super::GenericError("timed out waiting for response")))
+            }
+        }
+    }
+
+    fn send(&mut self, request: Request) -> Result<(), Box<dyn std::error::Error>> {
+        if let StreamState::Waiting(recv) = &self.stream {
+            if let Ok(stream) = recv.try_recv() {
+                self.stream = StreamState::Connected(stream);
+            }
+        }
+
+        match &mut self.stream {
+            StreamState::Connected(stream) => {
+                let data = bincode::serialize(&request)?;
+                stream.write_all(&(data.len() as u32).to_le_bytes())?;
+                stream.write_all(&data[..])?;
+                stream.flush()?;
+                Ok(())
+            }
+
+            _ => {
+                // Success if not connected (kinda dumb)
+                Ok(())
+            }
+        }
+    }
+
+    fn send_or_disconnect(&mut self, request: Request) -> Result<(), Box<dyn std::error::Error>>  {
+        if let Err(e) = self.send(request) {
+            self.disconnect();
+            return Err(e);
+        }
+
+        Ok(())
+    }
+
+    pub fn disconnect(&mut self) {
+        debug_output!(in self.seq, "[auxtools] disconnecting");
+        let _ = self.send(Request::Disconnect);
+
+        if let StreamState::Connected(stream) = &self.stream {
+            let _ = stream.shutdown(std::net::Shutdown::Both);
+        }
+
+        self.stream = StreamState::Disconnected;
+    }
+
+    pub fn get_line_number(&mut self, path: &str, override_id: u32, offset: u32) -> Result<Option<u32>, Box<dyn std::error::Error>> {
+        self.send_or_disconnect(Request::LineNumber {
+            proc: ProcRef {
+                path: path.to_owned(),
+                override_id,
+            },
+            offset,
+        })?;
+
+        match self.read_response_or_disconnect()? {
+            Response::LineNumber { line } => Ok(line),
+            _ => Err(Box::new(super::GenericError("received incorrect response"))),
```

Comment:
These error messages should contain at least what was expected and what they got instead

Reply:
Output is now like:
`[main] Error responding to "threads": received unexpected response: expected Stacks, got Offset { offset: Some(112) }`